### ISSUE TITLE
Vue warns about missing required prop that has a default value

### ISSUE THREAD

**Issue by decademoon** [vuejs/vue#7720]
### Version
2.5.13

### Reproduction link
[http://jsfiddle.net/df4Lnuw6/207/](http://jsfiddle.net/df4Lnuw6/207/)

### Steps to reproduce
Specify a required prop with a default value:

```js
Vue.component('comp', {
    template: '<div>{{ typeof x }} {{ x }}</div>',
    props: {
        x: {
            type: Number,
            required: true,
            default: 5,
        },
    },
});
```

Render the component without specifying a value for that prop:

```html
<comp></comp>
```

### What is expected?
The component should render the following without any prop validation errors:

```html
<div>number 5</div>
```

### What is actually happening?
The component renders OK, but warns about missing required prop `x`.

---
While it's true that prop `x` is not specified, since it has a default value, there should be no warning message.

What exactly does `required` check for? It appears that it checks two things:

1. The prop should be *provided*, as in at least `<comp :x="..."></comp>`.
2. The prop value should be non-null and non-undefined.

I think in the case where a required prop has a default value, (1) should be relaxed.

Otherwise, how can I enforce a prop to never be null while also providing a default value if the prop value was not provided?

**Comment by posva** [comment]
TLDR; Maybe we should indeed allow required + default value in [this check](https://github.com/vuejs/vue/blob/dev/src/core/util/props.js#L101)

By saying the prop is required you expect the user to always provide a value, so making it have a default value defeats the purpose of the prop being required.
When you require a prop, it means that you need that information from the user and that there is no way for you to provide a fallback value.

> Otherwise, how can I enforce a prop to never be null while also providing a default value if the prop value was not provided?

No, because `null` and `undefined` values always skip validation

If you want to take a deeper look at how the prop validation works, you can check https://github.com/vuejs/vue/blob/dev/src/core/util/props.js#L94

After doing some research in past issues, I found [this comment](https://github.com/vuejs/vue/issues/6768#issuecomment-335527318)

> Hmm, yeah so it needs to be more accurate:
> 
> - `null` indicates the value is explicitly marked as not present and it should remain `null`.
> - `undefined` indicates the value is not present and a default value should be used if available.
> - `required: true` indicates neither `null` or `undefined` are allowed (unless a default is used)
> 
> I agree this could be confusing, just as the existence of `null` vs. `undefined` in the language itself. But the design was trying to stay close to what these values are designed represent in JS, and changing them would be break

**Comment by decademoon** [comment]
Yeah, currently the issue that is `required: true` implies that the prop cannot be absent, but then if `required: false` and the prop value is nully then validation is skipped altogether. There is no way to enforce a prop to be non-nully at all times without `required: true` and hence prop omission is disallowed.

I also think the docs could be more clear about the precise definition of the `required` option.

**Comment by posva** [comment]
The validation will always be skipped with nully values though.
Contributions to docs are always welcome 😄

**Comment by javoski** [comment]
>The validation will always be skipped with nully values though.

Seems NOT when `required: true` is given. [http://jsfiddle.net/df4Lnuw6/225/](http://jsfiddle.net/df4Lnuw6/225/)

<img width="535" alt="2018-02-28 7 12 42" src="https://user-images.githubusercontent.com/5391892/36784724-71fd92dc-1cbb-11e8-978a-de94a2acec20.png">

**Comment by posva** [comment]
That's right! it is for undefined that validation is skipped with non-required props

**Comment by yyx990803** [comment]
How about a new flag `useDefaultForNull`

``` js
Vue.component('comp', {
    template: '<div>{{ typeof x }} {{ x }}</div>',
    props: {
        x: {
            type: Number,
            default: 5,
            useDefaultForNull: true
        },
    },
});
```

**Comment by Paquette** [comment]
It also reads ... awkwardly. Am I the only one who thinks that? Personally I'd go with adding `nullable` as a `type` declaration and using that as a flag, but I realize that's a little outside of the typical Vue approach.

**Comment by verymuch** [comment]
> By saying the prop is required you expect the user to always provide a value, so making it have a default value defeats the purpose of the prop being required.

Actually, `required: ture` and default value are coexisting. If the value is `undefined`, the default value will apply.

**Comment by DarioDaF** [comment]
> > By saying the prop is required you expect the user to always provide a value, so making it have a default value defeats the purpose of the prop being required.
> 
> Actually, `required: ture` and default value are coexisting. If the value is `undefined`, the default value will apply.

In fact it works as intended, it just warns for no reason (since property IS present cause not null in code being it the default value)

**Comment by moritz-baecker-integra** [comment]
Is there any new information on this issue?
We got the same issue with vue spamming the developer console with "missing prop" even though it has a default. 
Maybe i understood this wrong and there is a solution for this?

**Comment by albertAtPropertyMe** [comment]
> Is there any new information on this issue? We got the same issue with vue spamming the developer console with "missing prop" even though it has a default. Maybe i understood this wrong and there is a solution for this?

Perhaps. My recent experience with this has lead to some thoughts on this issue that I think could be helpful.

## Composition vs. Options

In our use case the confusion seems to arise from because of the differing defaults of composition API's `setup` script and options API `required: boolean` option. 

We use a combination of `defineProps` and `withDefaults` like so with composition API and Typescript,

```js
// for example
export interface Props {
  label: string
  isDisabled: boolean
}

const props = withDefaults(defineProps<Props>(), {
  isDisabled: false
});
```

Did you see it? The typo was as easy as missing an `?` on the optional prop when defining the props,

```js
// fixes the warning
export interface Props {
  label: string
  isDisabled?: boolean
}
```

Which is reasonably easy to miss, especially since you might have read this line in the docs [here](https://vuejs.org/guide/components/props.html#prop-validation) (can you forgive me for skim reading?),

![image](https://github.com/user-attachments/assets/4906e4d7-d04e-4265-8395-664293f6a582)

which would mean (like silly me) you would've missed this segment farther down,

![image](https://github.com/user-attachments/assets/e0422e15-f02f-420d-97da-44cd060c2318)

> In fact it works as intended, it just warns for no reason (since property IS present cause not null in code being it the default value)

Yes it does, however I might be forgiven for quickly parsing the docs and making the wrong assumption. 

## Assumptions about `withDefaults`

I would also point to this page [here](https://vuejs.org/guide/typescript/composition-api#props-default-values) that seems to indicate the opposite unless I'm completely misreading it,

![image](https://github.com/user-attachments/assets/127c1d53-c410-4567-9a17-2cfae8f1b22e)

> This will be compiled to equivalent runtime props default options. In addition, the withDefaults helper provides type checks for the default values, **and ensures the returned props type has the optional flags removed for properties that do have default values declared**.

## What could be improved here?

Commentary without suggestion isn't very helpful in my mind - so here's two things I would like to see improved,

1. This [section](https://vuejs.org/guide/components/props.html#prop-validation) on prop validation should probably link to [this section](https://vuejs.org/guide/typescript/composition-api#props-default-values) on composition API's Typescript validation. These topics go hand in hand and they currently feel disjointed.
2. I feel the usage of `withDefaults` should automagically make the prop optional, as mentioned above, this would be my assumption reading about it in the docs


The former seems quite simple and I'd happily take a look at contributing to the docs if that seems acceptable.

The latter is more opinionated though but at the very least I think the docs could be less confusing. Personally, I would expect that using `withDefaults` would implicitly define the provided props **as being optional**. Otherwise what's the point? I find this the exact kind of footgun that creates confusion in the first place, as it requires me to remember to specify optionality in two separate places (the prop definition as well as the default hook call) - _but this is purely my opinion_.
